### PR TITLE
Add delete priority annotation for ReplicaSet 

### LINKED ISSUE #45509: Scale down a deployment by removing specific pods
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): deployment scale down pod

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

It would be convenient to be able to scale down a deployment by N replicas by choosing which pods to remove from the deployment. 

Ideally, this would have a similar interface as for removing selected instances from a GCE managed instance group (https://cloud.google.com/sdk/gcloud/reference/compute/instance-groups/managed/delete-instances). 

Today, you can approximate this by issuing API calls in rapid succession that might sometimes result in that effect (remove the relevant label from the pod in question, scale the deployment and its underlying replicaSet down, then delete the pod), but it's hacky and relies on winning a race.

@kubernetes/sig-apps-feature-requests

### PR DESCRIPTION
**Pull request by drbugfinder**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

For some applications it is necessary that the application can tell Kubernetes which pod can be deleted and which replica has to be protected. The reason for this is that some applications do have stateful sessions and it is not possible to put such an application into Kubernetes because of data or session corruption resulting from "random" down-scale. If the application is able to tell Kubernetes which of the replicas contains no/few/less important active sessions, this would solve many problems. This PR is non-disruptive to the default behavior. Only if the annotation is existing, it will make a difference in deletion order.

The original PR (https://github.com/kubernetes/kubernetes/pull/75763) was closed and contained a coding error in the sorting function: If one pod was annotated and one was not, it could happen that the annotated pod gets deleted but the not annotated (but less prioritized) one stayed. This issue is fixed in this PR.

**Which issue(s) this PR fixes**:

Fixes #45509

**Special notes for your reviewer**:
For demo see my screencast: https://youtu.be/91APTsEsxIU

**Does this PR introduce a user-facing change?**:
A pod can be annotated with a controller.alpha.kubernetes.io/delete-priority annotation. If the annotation is existing, a high number (low priority) gets deleted first. If one replica does not have this annotation, it gets deleted first. 
This PR DOES NOT change any default behavior. Only if the annotation is existing, it will make a difference in deletion order.

### CODE REVIEW

File: pkg/controller/controller_utils.go [hunk 1]
```diff
@@ -51,6 +51,7 @@ import (
 	"k8s.io/utils/integer"
 
 	"k8s.io/klog"
+	"strconv"
```
Comment (skilxn-go) [inline pkg/controller/controller_utils.go]: Wrong order of headers.

File: pkg/controller/controller_utils.go [hunk 2]
```diff
@@ -81,6 +82,7 @@ const (
 	//      1+floor(log_2(ceil(N/SlowStartInitialBatchSize)))
 	SlowStartInitialBatchSize = 1
 )
+const DeletePriorityPodAnnotationKey = "controller.alpha.kubernetes.io/delete-priority"
```
Comment (skilxn-go) [inline pkg/controller/controller_utils.go]: If this value is not used out of this module, how about using lower case?

Comment (drbugfinder) [reply pkg/controller/controller_utils.go]: This part comes from the original commit https://github.com/kubernetes/kubernetes/pull/75763/commits
Feel free to make suggestions.

File: pkg/controller/controller_utils.go (hunk 1 above)
Comment (drbugfinder) [reply pkg/controller/controller_utils.go]: This part comes from the original commit https://github.com/kubernetes/kubernetes/pull/75763/commits

**Comment by neolit123** [comment]
@drbugfinder 
thank you for the PR!

please sign the CLA as the bot instructed above, because without the CLA check the PR cannot be accepted. thank you!

> do-not-merge/release-note-label-needed

the PR template (present in .github/) includes a "user facing change section" that needs to be present in the PR. from there a release note is picked up and this label will be removed.

**Pull request closed without merging**